File: AvlTree.java

class AvlNode
    {
            // Constructors
        AvlNode( Comparable theElement )
        {
            this( theElement, null, null );
        }

        AvlNode( Comparable theElement, AvlNode lt, AvlNode rt )
        {
            element  = theElement;
            left     = lt;
            right    = rt;
            height   = 0;
        }
        Comparable element;      
        AvlNode    left;         
        AvlNode    right;        
        int        height;       
    }
    public class AvlTree
    {
        public AvlTree( )
        {
            root = null;
        }

        public void insert( Comparable x )
        {
            root = insert( x, root );
        }

        public Comparable find( Comparable x )
        {
            return elementAt( find( x, root ) );
        }

        public void makeEmpty( )
        {
            root = null;
        }

        public boolean isEmpty( )
        {
            return root == null;
        }

        public void printTree( )
        {
            if( isEmpty( ) )
                System.out.println( "Empty tree" );
            else
                printTree( root );
        }

        private Comparable elementAt( AvlNode t )
        {
            return t == null ? null : t.element;
        }

        private AvlNode insert( Comparable x, AvlNode t )
        {
            if( t == null )
                t = new AvlNode( x, null, null );
            else if( x.compareTo( t.element ) < 0 )
            {
                t.left = insert( x, t.left );
                if( height( t.left ) - height( t.right ) == 2 )
                    if( x.compareTo( t.left.element ) < 0 )
                        t = rotateWithLeftChild( t );
                    else
                        t = doubleWithLeftChild( t );
            }
            else if( x.compareTo( t.element ) > 0 )
            {
                t.right = insert( x, t.right );
                if( height( t.right ) - height( t.left ) == 2 )
                    if( x.compareTo( t.right.element ) > 0 )
                        t = rotateWithRightChild( t );
                    else
                        t = doubleWithRightChild( t );
            }
            else
                ;  // Duplicate; do nothing
            t.height = max( height( t.left ), height( t.right ) ) + 1;
            return t;
        }

        private AvlNode find( Comparable x, AvlNode t )
        {
            while( t != null )
                if( x.compareTo( t.element ) < 0 )
                    t = t.left;
                else if( x.compareTo( t.element ) > 0 )
                    t = t.right;
                else
                    return t;    // Match

            return null;   // No match
        }

        private void printTree( AvlNode t )
        {
            if( t != null )
            {
                printTree( t.left );
                System.out.println( t.element );
                printTree( t.right );
            }
        }

        private static int height( AvlNode t )
        {
            return t == null ? -1 : t.height;
        }

        private static int max( int lhs, int rhs )
        {
            return lhs > rhs ? lhs : rhs;
        }

        private static AvlNode rotateWithLeftChild( AvlNode k2 )
        {
            AvlNode k1 = k2.left;
            k2.left = k1.right;
            k1.right = k2;
            k2.height = max( height( k2.left ), height( k2.right ) ) + 1;
            k1.height = max( height( k1.left ), k2.height ) + 1;
            return k1;
        }

        private static AvlNode rotateWithRightChild( AvlNode k1 )
        {
            AvlNode k2 = k1.right;
            k1.right = k2.left;
            k2.left = k1;
            k1.height = max( height( k1.left ), height( k1.right ) ) + 1;
            k2.height = max( height( k2.right ), k1.height ) + 1;
            return k2;
    	}

        private static AvlNode doubleWithLeftChild( AvlNode k3 )
        {
            k3.left = rotateWithRightChild( k3.left );
            return rotateWithLeftChild( k3 );
        }

        private static AvlNode doubleWithRightChild( AvlNode k1 )
        {
            k1.right = rotateWithLeftChild( k1.right );
            return rotateWithRightChild( k1 );
        }

        private AvlNode root;


        public static void main( String [ ] args )
        {
		AvlTree t = new AvlTree( );
		while(true)
		{
			System.out.println("1:Isert\n2:Find\n3:print\n4:Exit\tInsert choice\n");
			int ch = Integer.parseInt(System.console().readLine());
			int num;
			switch(ch) {
				case 1: System.out.println("Enter number\n");
					 num = Integer.parseInt(System.console().readLine());
					 t.insert(num);
					 break;
				case 2:  System.out.println("Enter number\n");
					 num = Integer.parseInt(System.console().readLine());
					 t.find(num);
					 break;
				case 3: t.printTree();
					break;
				case 4:	return;
				default: return;
			}
		}
	}
 }

File: avltree.c
#include<stdio.h>
#include<stdlib.h>
#include<malloc.h>

struct treenode {
	int element;
	struct treenode *left;
	struct treenode *right;
	int height;
};

int height(struct treenode *p)
{
	if(p == NULL)
		return -1;
	else
		return p->height;
}

/*Initialize the treenode*/
struct treenode *makenode(struct treenode *p, int data)
{
	p = (struct treenode *) malloc(sizeof (struct treenode));
	if(p == NULL)
		printf("Out of space\n");
	else {
		p -> element = data;
		p -> right = p -> left = NULL;
		p -> height = 0;
	}
	return p;
}

struct treenode *empty(struct treenode *p)
{
	if(p != NULL){
		p -> right = p -> left = NULL;
		free(p);
	}
	return NULL;
}


/*display in post order*/
void display(struct treenode *p)
{
	if(p -> left != NULL)
		display(p -> left);
	if(p -> right != NULL)
		display(p -> right);
	printf("%d\t", p -> element);
}
/*return greatest number*/
int max(int a,int b)						
{
	return a > b ? a : b;
}

/*right rotation */
struct treenode *single_rightrotation(struct treenode *p)	
{
	struct treenode *pivot;
	pivot = p -> left;
	p -> left = pivot -> right;
	pivot -> right = p;

	p -> height = max(height(p -> left), height(p -> right)) + 1;
	pivot -> height = max(height(pivot -> left), height(pivot -> right)) + 1;
	return pivot;						//This will become our new p
}

/*left rotation*/
struct treenode *single_leftrotation(struct treenode *p)
{
	struct treenode *pivot = p -> right;
	p -> right = pivot -> left;
	pivot -> left = p;

	p -> height = max(height(p -> left), height(p -> right)) + 1;
	pivot -> height = max(height(pivot -> left), height(pivot -> right)) + 1;
	return pivot;						//This will become our new p
}

/*left-right case
  **The first rotation is a left rotation with L as the root. The second is a right rotation with P as the root
 */
struct treenode *doublerotation_withleft(struct treenode *p)
{
	p -> left = single_leftrotation(p -> left);
	return single_rightrotation(p);
}

/* right-left case
   The first rotation is a right rotation with R as the root. The second is a left rotation with P as the root */
struct treenode *doublerotation_withright(struct treenode *p)
{
	p -> right = single_rightrotation(p -> right);
	return single_leftrotation(p);
}

struct treenode *insert(struct treenode *p, int data)
{
	if(p == NULL)
		p = makenode(p,data);
	else {
		if(data < p -> element) {
			p -> left = insert(p -> left, data);
			if((height(p -> left) - height(p -> right)) == 2)
				if((height(p -> left -> left) - height(p -> left ->right)) == 1)
					p = single_rightrotation(p);		//single right rotation with p as root is needed
				else
					p = doublerotation_withleft(p);		//left-right case
		}
		else {
			if(data > p ->element){
				p -> right = insert(p -> right, data);
				if((height(p -> left) - height(p -> right)) == -2)
					if((height(p -> right -> left) - height(p -> right ->right)) == -1)
						p = single_leftrotation(p);	//a single left rotation (with P as the root) is needed (Right-Right case).
				        else
					        p = doublerotation_withright(p);	//right- left case										
			}
		}
	}
	p -> height = max(height(p ->right), height(p -> left)) + 1;
	return p;
}

struct treenode *lookup(struct treenode *p, int data)
{
	if(p == NULL)
		return NULL;
	if(data < p -> element)
		return lookup(p -> left,data);
	else if(data > p -> element)
		return lookup(p -> right,data);
	else if(data == p -> element)
		return p;
	else 
		return NULL;
}


int main()
{
	struct treenode *t;
	int element, choice;
	t = empty(NULL);
	while(1){
		printf("\nEnter your choice\t1:insert\t2:lookup\t3:display\t4:exit\t");
		scanf("%d",&choice);
		switch(choice){
			case 1: printf("Enter the element\n");
				scanf("%d",&element);
				t = insert(t, element);
				break;
			case 2: printf("Enter the element\n");
				scanf("%d",&element);
				if(lookup(t, element))
					printf("Found!!\n");
				else
					printf("Not found\n");
				break;
			case 3: printf("Printing in postorder\n");
				display(t);
				break;
			case 4 :exit(0);
			default: printf("Wrong choice\n");
		}
	}
	return 0;
}

